mouse.go

```diff
@@ -58,45 +58,62 @@ var mouseEventTypes = map[MouseEventType]string{
 //
 //     ESC [M Cb Cx Cy
 //
+// See: http://www.xfree86.org/current/ctlseqs.html#Mouse%20Tracking
 func parseX10MouseEvent(buf []byte) (m MouseEvent, err error) {
 	if len(buf) != 6 || string(buf[:3]) != "\x1b[M" {
 		return m, errors.New("not an X10 mouse event")
 	}
 
 	e := buf[3] - 32
 
-	switch e {
-	case 35:
-		m.Type = MouseMotion
-	case 64:
-		m.Type = MouseWheelUp
-	case 65:
-		m.Type = MouseWheelDown
-	default:
-		switch e & 3 {
-		case 0:
-			if e&64 != 0 {
-				m.Type = MouseWheelUp
-			} else {
-				m.Type = MouseLeft
-			}
-		case 1:
-			if e&64 != 0 {
-				m.Type = MouseWheelDown
-			} else {
-				m.Type = MouseMiddle
-			}
-		case 2:
+	const (
+		bitShift  = 0b0000_0100
+		bitAlt    = 0b0000_1000
+		bitCtrl   = 0b0001_0000
+		bitButton = 0b0010_0000
+		bitWheel  = 0b0100_0000
+
+		bitsRelease = 0b0000_0011
+
+		bitsMask = 0b0000_0011
+
+		bitsLeft   = 0b0000_0000
+		bitsMiddle = 0b0000_0001
+		bitsRight  = 0b0000_0010
+		bitsMotion = 0b0000_0011
```

Comment:
I _think_ the motion bits here are actually `0b0010_0011` (was `35` as an integer in the previous version of the code).

Reply:
Yeh, but we've already checked 6'th bit by `e&bitButton`:

```
bitButton | bitsMotion = 0b0010_0000 | 0b0000_0011 = 0b0010_0011

```py
# For example right click (0b0010_0010).
# We check this is the button event.
0b0010_0010 & bitButton = 0b0010_0010 & 0b0010_0000 = 0b0010_0000

# Then mask the low two bits.
0b0010_0010 & bitsMask = 0b0010_0010 & 0b0000_0011 = 0b0000_0010

# Then check this bits.
0b0000_0010 == bitsRight = 0b0000_0010 == 0b0000_0010 = true
```

The same with scroll events. I checked all events in tests.

Reply:
You're right again :). Makes sense.